### ISSUE TITLE
libclang version does not match Clang's

### ISSUE THREAD

**Issue by w62** [Rust-for-Linux/linux#777]
Hi Miguel and the seasoned developers,

I have getting the following message:

```
[william@localhost linux]$ make LLVM=1 rustavailable
***
*** libclang (used by the Rust bindings generator 'bindgen') version does not match Clang's. This may be a problem.
***   libclang version: 11.0.0
11.0.0
***   Clang version:    11.0.0
***
Rust is available!
[william@localhost linux]$

```

It seems that 11.0.0 is not equal to 11.0.0. I don't know what I have dnf installed llvm and it does not help. When I do make menuconfig, Rust does not appear under General Setup. In fact I have tried the following combination and none of them work. 

- ubuntu 20.04 aarch64
- ubuntu 20.04 x86_64

- fedora 36 aarch64
- fedora 36 x86_64

- fedora 34 aarch64

My last test is fedora 33 x86_64. I ended up this version because it is the version with default clang 11.

I am following the [the quick start guide](https://github.com/Rust-for-Linux/linux/blob/rust/Documentation/rust/quick-start.rst). 

For the aarch64, I am using VM under Parallel in a M1 mac. For the x86_64, I am using VM under VirtualBox in a Intel macbook air.

Does the project works under aarch64 currently? If so, please share with me the detailed procedure. I am really interested in getting Rust to work with the Linux kernel.

thanks

william

Edit: It actually should be "clang does not match libclang" in the title.

**Comment by ojeda** [comment]
> It seems that 11.0.0 is not equal to 11.0.0

Hmm... In the output there are two consecutive `11.0.0` strings separated by a newline apparently, which would make them non-equal. Could you please show the output of:

```sh
LC_ALL=C bindgen scripts/rust-is-available-bindgen-libclang.h
```

and of:

```sh
LC_ALL=C clang --version
```

?

> Does the project works under aarch64 currently?

Yes, it should (we boot test it in the CI). The host architecture should not really matter, though. As long as you are compiling under Linux itself, it should be fine.

**Comment by ojeda** [comment]
By the way:

> When I do make menuconfig, Rust does not appear under General Setup

Note that if `rustavailable` was successful (as you showed above -- that message is a warning, not an error), then Rust should appear in that menu as long as the other prerequisites are satisfied. Could you please verify that is the case? (e.g. check using the search command `/` inside `menuconfig`).

**Comment by w62** [comment]
Thanks for looking into the issue and helping me out.

Please find the output:

```
[william@hats linux]$ LC_ALL=C bindgen scripts/rust-is-available-bindgen-libclang.h
scripts/rust-is-available-bindgen-libclang.h:2:9: warning: clang version 11.0.0 (Fedora 11.0.0-3.fc33) [-W#pragma-messages], err: false
/* automatically generated by rust-bindgen 0.56.0 */


[william@hats linux]$
[william@hats linux]$ LC_ALL=C clang --version
clang version 11.0.0 (Fedora 11.0.0-3.fc33)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
[william@hats linux]$

```

The search in make menuconfig does show that presence of HAVE_RUST defaults to yes

```
Symbol: HAVE_RUST [=y]                                                                                   
  Type  : bool                                                                                             
   Defined at arch/Kconfig:347                                                                            
   Selected by [y]:                                                                                         
     - X86 [=y] && X86_64 [=y]
   ```

**Comment by w62** [comment]
I have run the same command again in my aarch64 Ubuntu. Please find the output:

```
09:11:06
[ 🐔  parallels@ubuntu-linux-20-04-desktop:/home/parallels/projects/linux]
:P > LC_ALL=C bindgen scripts/rust-is-available-bindgen-libclang.h
scripts/rust-is-available-bindgen-libclang.h:2:9: warning: clang version 11.0.0  [-W#pragma-messages], err: false
/* automatically generated by rust-bindgen 0.56.0 */



09:11:30
[ 🐔  parallels@ubuntu-linux-20-04-desktop:/home/parallels/projects/linux]
:P > LC_ALL=C clang --version
Ubuntu clang version 11.0.0-2~ubuntu20.04.1
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

09:11:54
[ 🐔  parallels@ubuntu-linux-20-04-desktop:/home/parallels/projects/linux]
:P >

```

make menuconfig also shows HAVE_RUST defaults to yes
```
  Symbol: HAVE_RUST [=y]                                                                                               
   Type  : bool                                                                                                         
   Defined at arch/Kconfig:347                                                                                          
   Selected by [y]:                                                                                                     
     - ARM64 [=y]
```   

So far I am using a non-root user to do it in virtual machines. 

Is that possible for you to share your development environment or how to set it up?

thanks again.

**Comment by ojeda** [comment]
> ```
> scripts/rust-is-available-bindgen-libclang.h:2:9: warning: clang version 11.0.0 (Fedora 11.0.0-3.fc33) [-W#pragma-messages], err: false
> ```

This is where the two `11.0.0` are being picked up -- thanks! I will fix it so that you don't see the warning.

> The search in make menuconfig does show that presence of HAVE_RUST defaults to yes
> 
> ```
> Symbol: HAVE_RUST [=y]                                                                                   
>   Type  : bool                                                                                             
>    Defined at arch/Kconfig:347                                                                            
>    Selected by [y]:                                                                                         
>      - X86 [=y] && X86_64 [=y]
> ```

This means you have the toolchain available, which is what the script showed, so this is correct and expected.

Instead, please take a look at `RUST`, not `HAVE_RUST`. That will show you why the option cannot be enabled.

**Comment by w62** [comment]
Could you teach me how to fix the script or where to start with? I want to take this opportunity to understand the project more.

I have recorded my make menuconfig session [here](https://youtu.be/brQz6g16OTw).  Rust hacking indeed show up, but there left arrow '>' is missing and there is no option when I entered it.

**Comment by tiagolam** [comment]
Didn't notice there was an issue open about this already, and ended up addressing this in PR #789.

@w62 I already have a fix for this, but I'll dump here my rationale in the hope this provides some context for you, if you want to take a better look / contribute in the future.

As you and Miguel noticed in one of the comments above, calling `LC_ALL=C bindgen scripts/rust-is-available-bindgen-libclang.h` was giving the following result:
```
$ LC_ALL=C bindgen scripts/rust-is-available-bindgen-libclang.h
scripts/rust-is-available-bindgen-libclang.h:2:9: warning: clang version 13.0.0 (Fedora 13.0.0-3.fc35) [-W#pragma-messages], err: false
/* automatically generated by rust-bindgen 0.56.0 */
```

This was suspicious, given that, as you pointed out in your first message, the `13.0.0` was indeed being reported twice:
```
$ make LLVM=1 rustavailable
***
*** libclang (used by the Rust bindings generator 'bindgen') version does not match Clang's. This may be a problem.
***   libclang version: 13.0.0
13.0.0
***   Clang version:    13.0.0
***
Rust is available!
```

So, next step was to find what was calling `scripts/rust-is-available-bindgen-libclang.h` and check how the result being parsed. It turns out this is happening currently in `scripts/rust-is-available.sh`, here: https://github.com/Rust-for-Linux/linux/blob/9f4510ea769db8ea6d974f11a45322a1cf55e6ca/scripts/rust-is-available.sh#L104

We can now isolate that line an call it individually to check what's going on, and indeed we get:
```
$ LC_ALL=C bindgen ./scripts/rust-is-available-bindgen-libclang.h 2>&1 >/dev/null | grep -F 'clang version ' | grep -oE '[0-9]+\.[0-9]+\.[0-9]+'
13.0.0
13.0.0
```

And this where our double `13.0.0` is coming from, and why it isn't matching the installed Clang version, which reports a single `13.0.0`.

Given we only want the first line, we can use `head` here, which is commonly used for this sort of tasks (and even used in other parts of the rust-is-available.sh script, if you take a look). With this, we end up with:
```
$ LC_ALL=C bindgen ./scripts/rust-is-available-bindgen-libclang.h 2>&1 >/dev/null | grep -F 'clang version ' | grep -oE '[0-9]+\.[0-9]+\.[0-9]+' | head -1
13.0.0
```

Hope this helps giving some clarification.

**Comment by ojeda** [comment]
Should be fixed by #789 -- can you please give it a go @w62? If it does not work, we can reopen.

**Comment by w62** [comment]
Yes, it is fixed. Thanks for looking into the issue.
